src/vs/editor/common/modes.ts

```diff
@@ -392,7 +392,7 @@ export interface ReferenceProvider {
 	/**
 	 * Provide a set of project-wide references for the given position and document.
 	 */
-	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, token: CancellationToken): Location[] | Thenable<Location[]>;
+	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, progress: (value: Location) => void, token: CancellationToken): Location[] | Thenable<Location[]>;
```

Comment:
You are probably right about getting feedback early (and probably have more experience with that than I do).
But our implementation here should at least be forward-compatible / in-line with the LSP spec, eg we should at least have a way in our minds how we _would_ wire this up to a real LS that uses the streaming spec. Currently this is not possible, because language servers do not always emit full `Locations`. They emit operations that are applied the final result and these can contain arbitrary modifications. It is hard/impossible to extract full elements from these operations.

I will do the LSP proposal.